### ISSUE TITLE
lib/precode.c:121: precode_matrix_choose: Assertion `0' failed.

### ISSUE THREAD

**Issue by Intrepd** [sleepybishop/nanorq#13]
First, thank you for your work on the library.

I was trying to write a test case that covers recovery failure, and I received the assertion error in the title.  

In summary, using exactly 10 source symbols of 1448 bytes and dropping ESI 4, 5, 6, 10, and 11.

Per a suggestion in another issue, I was able to modify benchmark.c to reproduce this:
[benchmark.zip](https://github.com/sleepybishop/nanorq/files/10711230/benchmark.zip)

> ./benchmark 1448 10 50
adding symbol 0 to sbn 0.
adding symbol 1 to sbn 0.
adding symbol 2 to sbn 0.
adding symbol 3 to sbn 0.
adding symbol 7 to sbn 0.
adding symbol 8 to sbn 0.
adding symbol 9 to sbn 0.
adding symbol 12 to sbn 0.
adding symbol 13 to sbn 0.
adding symbol 14 to sbn 0.
benchmark: lib/precode.c:121: precode_matrix_choose: Assertion `0' failed.
Aborted (core dumped)

FWIW, the Rust RaptorQ implementation seems to be able to perform a successful recovery with this data set.

**Comment by sleepybishop** [comment]
Thank you for raising this.

The assert to is there to verify that section was unreachable after I removed some fallback code, and it's obviously not.

The assert is being triggered because there are no symbols that can be easily untangled and the fallback code was there choose the least tangled one.

That being said, the chooser logic in `nanorq` is deliberately quite different from the rfc and other implementations. In the `dev` branch I've even made it customizable as ultimately the application might need different implementations depending on the use case and device constraints.

I'll take a look at restoring the fallback code and take a closer look at the resulting decode matrix to make sure there is nothing else going on.

**Comment by sleepybishop** [comment]
I've restored the fallback code in e8d6c8fc3e0ab012926ab78eb9d5c40804488c7d and verified the attached test no longer fails.

**Comment by Intrepd** [comment]
Great, thanks.  I pulled it down an as you would expect the assertion error is gone.

Would you also mind running the attached benchmark code? [benchmark.zip](https://github.com/sleepybishop/nanorq/files/10739778/benchmark.zip)

The difference in this version, is that I drop symbols 0, 3, 10, 12, and 14.

```
>./benchmark 1448 10 50
num_packet: 10
encode num_sbn: 1 overhead: 50.00
adding symbol 1 to sbn 0.
adding symbol 2 to sbn 0.
adding symbol 4 to sbn 0.
adding symbol 5 to sbn 0.
adding symbol 6 to sbn 0.
adding symbol 7 to sbn 0.
adding symbol 8 to sbn 0.
adding symbol 9 to sbn 0.
adding symbol 11 to sbn 0.
adding symbol 13 to sbn 0.
*** stack smashing detected ***: terminated
Aborted (core dumped)

>Program received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140737351501632) at ./nptl/pthread_kill.c:44
44      ./nptl/pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737351501632) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737351501632) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737351501632, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff7dc1476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff7da77f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff7e086f6 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7f5a943 "*** %s ***: terminated\n") at ../sysdeps/posix/libc_fatal.c:155
#6  0x00007ffff7eb576a in __GI___fortify_fail (msg=msg@entry=0x7ffff7f5a92b "stack smashing detected") at ./debug/fortify_fail.c:26
#7  0x00007ffff7eb5736 in __stack_chk_fail () at ./debug/stack_chk_fail.c:24
#8  0x000055555555e5f6 in precode_matrix_swap_cols (V0=V0@entry=5, Vcols=Vcols@entry=11, S=S@entry=0x555555583500, A=<optimized out>, A=<optimized out>) at lib/precode.c:158
#9  0x000055555555e8bb in precode_matrix_precond (P=P@entry=0x555555575510, A=A@entry=0x555555581140, AT=AT@entry=0x555555585980, S=S@entry=0x555555583500) at lib/precode.c:199
#10 0x000055555555f361 in precode_matrix_invert (P=P@entry=0x555555575510, A=A@entry=0x555555581140) at lib/precode.c:360
#11 0x0000555555559eff in nanorq_repair_block (rq=0x5555555754a0, io=0x555555575440, sbn=<optimized out>) at lib/nanorq.c:619
#12 0x0000555555555d26 in decode ()
#13 0x00005555555560b7 in run ()
#14 0x000055555555633d in main ()
```

**Comment by sleepybishop** [comment]
I enabled building with the stack protector to replicate locally (58941940d17e167644665b0acadee5b99ab17d5e). The problem is present in the last example as well.

It's rooted in the same cause though. I've been ignorant of the zero overhead case because I always try to provide enough extra symbols to speed up decoding. In any case, in the zero overhead case, a chosen row might be dependent on multiple symbols, in which case extra work needs to be done. Fixed in (5613cc1ed9395ee11c74d73f60874e8ad0e9f105)

**Comment by Intrepd** [comment]
Thank you!  This is working much better.

Could you look at this case where symbols 4, 5, 10, 11, and 14 are dropped?

[benchmark.zip](https://github.com/sleepybishop/nanorq/files/10757496/benchmark.zip)

I see that nanorq_repair_block claims success, but I don't think the in and out buffer match.

> ./benchmark 1448 10 50
num_packet: 10
encode num_sbn: 1 overhead: 50.00
adding symbol 0 to sbn 0.
adding symbol 1 to sbn 0.
adding symbol 2 to sbn 0.
adding symbol 3 to sbn 0.
adding symbol 6 to sbn 0.
adding symbol 7 to sbn 0.
adding symbol 8 to sbn 0.
adding symbol 9 to sbn 0.
adding symbol 12 to sbn 0.
adding symbol 13 to sbn 0.
memcmp result: -85

From what I've found, these combinations of symbols received report success, but seem to fail to recover properly:

>[0, 1, 2, 3, 6, 7, 8, 9, 12, 13]
[0, 1, 2, 3, 6, 7, 8, 9, 13, 14]
[0, 1, 2, 3, 7, 8, 9, 12, 13, 14]
[0, 1, 2, 6, 7, 8, 9, 12, 13, 14]
[0, 1, 3, 6, 7, 8, 9, 12, 13, 14]
[0, 2, 3, 6, 7, 8, 9, 12, 13, 14]
[1, 2, 3, 6, 7, 8, 9, 12, 13, 14]

**Comment by sleepybishop** [comment]
Not sure what is going here, I'll have to dive in and take a closer look.

**Comment by sleepybishop** [comment]
Unsure if this issue is still relevant. I recently had to revisit the zero overhead case for a delay tolerant scenario. I pushed two changes, the latter of which I think should resolve this issue.

  - The first change was to patch the modifications I had made to address the crashes reported here. It was a mistake to modify the preconditioner, it should stay as simple as possible. Making it customizable is something I started in the `dev` branch but it could also be easily replaced in this branch.
  - The second change is that the preconditioner should also not decide inversion success. It's role is to reduce the work done by the solvers, if it's not able to do much, it should not abort and let the solvers decide whether inversion is possible or not, even if that means higher cpu/memory costs due to a larger U matrix.
